Merge `feature/circadia` into `develop`

ScalaChamp opened the pull request:
## Context

_Say a few words about the bigger picture beyond this PR. How do your changes improve life for our users? Link directly to the jira ticket (just type "DDP-..." and it'll automatically link) to help your teammates quickly get up to speed on the motivation for these changes._

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

YuriyPavlovBI committed:
Circadia initial commit

ScalaChamp committed:
DDP-5517 | implement prequal for Circadia

ScalaChamp committed:
implement transition workflows for prequal

ScalaChamp committed:
Merge pull request #768 from broadinstitute/DDP-5517_prequal

DDP-5517 | implement prequal for Circadia

ScalaChamp committed:
crc Consent halfway done

YuriyPavlovBI committed:
Study guid change

ScalaChamp committed:
shrink expression size in study.conf

aleksey-kuznetsov committed:
Add circadia auth0 config files & templates

ScalaChamp committed:
make circadia buildable with default consent

ScalaChamp committed:
merge auth0 stuff with backend stuff

ScalaChamp committed:
DDP-5519 | implement consent.conf for Circadia

ScalaChamp committed:
Merge pull request #778 from broadinstitute/DDP-5519_consent

DDP-5519 | implement consent.conf for Circadia

ScalaChamp committed:
incorporate naming conventions in prequal for Circadia

ScalaChamp committed:
fix minor things in prequal for Circadia

ScalaChamp committed:
incorporate naming conventions in consent for Circadia

ScalaChamp committed:
DDP-5520 | implement Health and Lifestyle Questionnaire for Circadia

aleksey-kuznetsov committed:
Circadia: auth0 login / signup page changes

aleksey-kuznetsov committed:
Merge pull request #783 from broadinstitute/feature/DDP-5518_setup-auth0

Circadia: auth0 login / signup page changes

aleksey-kuznetsov committed:
Circadia: auth0 password reset page fixes

ScalaChamp committed:
DDP-6133 | implement post-collection-questionnaire for Circadia

ScalaChamp committed:
Merge pull request #786 from broadinstitute/DDP-6133_post_collection_questionnaire

DDP-6133 | implement post-collection-questionnaire for Circadia

ScalaChamp committed:
DDP-6171 | implement morning-evening-questionnaire for Circadia

ScalaChamp committed:
Merge pull request #788 from broadinstitute/DDP-6171_morning_evening_questionnaire

DDP-6171 | implement morning-evening-questionnaire for Circadia

ScalaChamp committed:
implement Circadia PILOT prequal

aleksey-kuznetsov committed:
Circadia: add auth0 welcome email

aleksey-kuznetsov committed:
Circadia: change Auth0 welcome email "from" field to use value from Auth0 Deploy CLI config

aleksey-kuznetsov committed:
Merge pull request #825 from broadinstitute/feature/DDP-6164_welcome-email

DDP-6164: Auth0 welcome email

ScalaChamp committed:
fix bug for consent of Circadia, redirect eligible participants to auth0 (#829)

aleksey-kuznetsov committed:
Circadia: fix whitespaces in Auth0 welcome email template

ScalaChamp committed:
fix bugs for consent and post-collection questionnaire for Circadia (#831)

* fix bugs for consent and post-collection questionnaire for Circadia

* change html class for question-type block from paragraph to question-title

aleksey-kuznetsov committed:
Circadia: reorder fields on login page

aleksey-kuznetsov committed:
Circadia: minor fixes for Auth0 login page

aleksey-kuznetsov committed:
Merge branch 'feature/circadia' of github.com:broadinstitute/ddp-study-server into feature/circadia

ScalaChamp committed:
fix small issues in consent for Circadia (#835)

aleksey-kuznetsov committed:
Circadia: small fix for mobile devices

ScalaChamp committed:
DDP-6386 | implement Kit Return Attestation activity for PILOT Circadia (#836)

ScalaChamp committed:
DDP-6171 | add morning-evening-questionnaire for Circadia (#840)

ScalaChamp committed:
remove activity-wide validation for prequal, leave validation only for user-specific options (#841)

ScalaChamp committed:
fix date in consent for Circadia

ScalaChamp committed:
add workflow transition after consent submission

ScalaChamp committed:
DDP-6135 (#842)

| implement DLMO Scheduling Instructions for Circadia

aleksey-kuznetsov committed:
DDP-6170: Add "DLMO Instructions" activity

aleksey-kuznetsov committed:
Merge branch 'feature/circadia' into feature/DDP-6170_dlmo-instructions

# Conflicts:
#	study-builder/studies/circadia/study.conf

aleksey-kuznetsov committed:
Merge pull request #843 from broadinstitute/feature/DDP-6170_dlmo-instructions

DDP-6170: DLMO Instructions activity

ScalaChamp committed:
add summary text for DLMO Scheduling Instructions for Circadia

aleksey-kuznetsov committed:
Circadia: remove redundant logout url from Auth0 config

ScalaChamp committed:
Ddp 5521 dashboard summaries (#844)

* add summary texts for the existing activities for Circadia

* add blank activities with summary texts for Circadia

* modify trigger actions in order to create all the necessary activities after consent submission for Circadia

ScalaChamp committed:
DDP-6385 | implement Kit Receipt Attestation activity for Circadia (#847)

* DDP-6385 | implement Kit Receipt Attestation activity for Circadia

* fix typo in en.json

ScalaChamp committed:
DDP-6381 | implement Pre-Collection Attestation #1 for Circadia (#848)

ScalaChamp committed:
DDP-6386 | add text field for additional comment for Kit Return Attestation & remove duplicate questionStableId-s (#850)

ScalaChamp committed:
DDP-6383 | implement Pre Collection Attestation #2 for Circadia & remove duplicate questionStableId-s (#851)

ScalaChamp committed:
DDP-6382 | implement Post Collection Attestation #1 for Circadia (#853)

ScalaChamp committed:
DDP-6386 | fix bug, now user is redirected to Dashboard after submitting Kit Return Attestation (#854)

ScalaChamp committed:
DDP-6384 | implement Post Collection Attestation #2 for Circadia (#855)

ScalaChamp committed:
DDP-6488 | attach download link for consent PDF for Circadia

ScalaChamp committed:
DDP-6488 | display download message on every section

ScalaChamp committed:
DDP-6382-6384 | display post collection activities on dashboard (#874)

ScalaChamp committed:
add dashboard summaries for Post-Collection Attestations

ScalaChamp committed:
fix bug in morning evening questionnaire in Circadia

ScalaChamp committed:
DDP-6595 | implement Health and Lifestyle Questionnaire for Circadia (#890)

* DDP-6595 | implement Health and Lifestyle Questionnaire for Circadia

* DDP-6595 | fix typo

* DDP-6595 | fix typo #2

ScalaChamp committed:
DDP-6601 | implement fixes from customer feedback (#895)

* DDP-6601 | implement fixes from customer feedback

* DDP-6601 | remove time field from Consent

ScalaChamp committed:
DDP-6597 | Update activity timing and dashboard summaries for Circadia

ScalaChamp committed:
Merge branch 'develop' into feature/circadia

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/model/dsm/DsmNotificationEventType.java:
```diff
@@ -33,4 +33,12 @@
+    /**
+     * Message sent by DSM when a Circadia kit is delivered to participant
+     */
+    CIRCADIA_SENT,
+    /**
+     * Message sent by DSM when a Circadia  kit is received back
+     */
+    CIRCADIA_RECEIVED
```
Can you point me at the ticket for this change? Does @SimoneMaiwald or @MocanaAtBroad know about these new notification types we're adding to the platform?

ScalaChamp replied:
I think it's this one:

https://broadinstitute.atlassian.net/browse/DDP-6574

yufengwng replied:
I'm not sure I like that we're adding notification types that's so study-specific (or at least, is named after the study). I know we did it as a special case for TestBoston, but still... What do you think @MocanaAtBroad?

ScalaChamp replied:
@yufengwng what if we renamed them as `KIT RECEIVED` and `KIT SENT`? it'll be more generic for the future events as well, @SimoneMaiwald what do you think?

yufengwng replied:
In the ticket, I see the original idea was to use `KIT_SENT` and `KIT_RECEIVED`. I'm curious why we decided to rename it to `CIRCADIA_*`? Also, how should we interpret the `KIT_*` types? How does it interact with other types like `BLOOD_SENT` and `SALIVA_RECEIVED`? Do we get both `KIT_SENT` and `BLOOD_SENT` when a blood kit is sent?

From what I can tell, these notification types are pretty isolated in DSS -- they don't really affect other features in the platform. So having something like `CIRCADIA_SENT` is fine in a sense. But I urge us to think a bit more about these since there's other studies we support and we'll need to document and explain why we have things like `CIRCADIA_SENT` in the platform.

ScalaChamp replied:
I was using `CIRCADIA_SENT` and `CIRCADIA_RECEIVED` because they were provided in the comments of the ticket, we can return those to the original values, so that they can be `KIT_SENT` and `KIT_RECEIVED`. I'm not sure about the reason of renaming to be honest, I was just following the requested solution.

The main reason of having these notifications is that they are used as `triggers` to create some activities (e.g in this case `DLMO Instructions` and `Kit Receipt Attestation`), also they are used in `emailing` too.

I don't know whether these notification types interact with other types in any way, for Circadia we just needed special triggers from DSM to control workflows and statuses.

I'm open to be included in more detailed discussion if it's necessary, maybe a special meeting with others too.

ScalaChamp replied:
Actually the real `trigger` is specified as this: `As soon as Kit Shipped Date is entered in DSM` and I'm note sure how it should happen in DSM since I'm not familiar with that project and also I'm not sure whether we have a similar `trigger` (Might be `BLOOD_SENT`) in DSS side.

We were using `KIT_SENT` and `KIT_RECEIVED` just to start out with that problem

yufengwng commented on study-builder/tenants/circadia/tenant.yaml:
```diff
@@ -0,0 +1,50 @@
+tenant:
+  default_directory: Username-Password-Authentication
+
```
Is the tenant created already? Can you make sure at the very least @MocanaAtBroad is added as an admin for the tenant?

aleksey-kuznetsov replied:
@yufengwng this tenant is already set up. I'll add Marco.

ScalaChamp reviewed (commented):
Yes, Simone is aware of these changes

yufengwng commented on study-builder/tenants/circadia/emails/welcome_email.html:
```diff
@@ -0,0 +1,146 @@
+</head>
+<body>
+<!-- Variable definitions -->
+{% assign logo_url = "https://storage.googleapis.com/circadia-dev-assets/auth0/logo.png" %}
```
I recommend pulling out `circadia-dev-assets` into a substitution. Add it to Vault and render it in `circadia-config.json.ctmpl`. Or maybe we can use `##ASSETS_URL##` like how we have it in `password_reset.html`?

aleksey-kuznetsov replied:
@yufengwng thank you, I'll handle it.

yufengwng commented on study-builder/tenants/circadia/pages/password_reset.html:
```diff
@@ -0,0 +1,191 @@
+      var study_guid = getQueryParamValue('study');
+      var studyColor = '#0b1245'; // button primary color
+      var headerText = '';
+      var headerLogo = '##ASSETS_URL##/auth0/logo.png';
```
Where is `ASSETS_URL` coming from? I don't see it in `circadia-config.json.ctmpl`?

yufengwng commented on study-builder/tenants/circadia/pages/login.html:
```diff
@@ -0,0 +1,546 @@
+
+      config.extraParams = config.extraParams || {};
+
+      var inviteKey = 'invitationId';
+
+      // save the invitation in local storage to minimize breached password loss-of-invitation
+      if (config.extraParams.invitation_id) {
+        localStorage.setItem(inviteKey, config.extraParams.invitation_id);
+
+        console.log('Saving invitation ' + config.extraParams.invitation_id);
+      } else {
+        var invitation = localStorage.getItem(inviteKey);
+
+        if (invitation) {
+          // unsafe, but better than losing this value during breached password shield
+          config.internalOptions.invitation_id = invitation;
+
+          console.log(
+            'Using stored invitation ' + invitation + ' from local storage',
+          );
+        }
+      }
```
Is Circadia using invitations? I don't think `invitation_id` is relevant and we should remove this block of code.

yufengwng commented on study-builder/tenants/circadia/pages/login.html:
```diff
@@ -0,0 +1,546 @@
+      var headerLogo = '##ASSETS_URL##/auth0/logo.png';
+      var headerText = '';
+      var studyColor = '#0b1245';
+      var studyHelpEmail = '##STUDY_EMAIL##';
```
I don't see `STUDY_EMAIL` being provided in `AUTH0_KEYWORD_REPLACE_MAPPINGS` in `circadia-config.json.ctmpl`.

yufengwng commented on study-builder/tenants/circadia/tenant.yaml:
```diff
@@ -0,0 +1,50 @@
+  - template: welcome_email
+    enabled: true
+    syntax: liquid
+    from: '##FROM_EMAIL##'
```
I don't see `FROM_EMAIL`... if we add `STUDY_EMAIL` maybe we can use that here? Also, I think the format for this `from` value looks something like
```
Study Name <email_here>
```

yufengwng commented on study-builder/tenants/circadia/tenant.yaml:
```diff
@@ -0,0 +1,50 @@
+    body: emails/welcome_email.html
+
+clients:
+  - name: Circadia Dev
```
We're going to run this `tenant.yaml` for all the different environments (including production!) to setup Auth0. I recommend naming this client `circadia-angular-client` or similar.

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+  "study": {
+    "name": "circadia",
+    "guid": "circadia",
+    "studyEmail": "cancergenomecollective@broadinstitute.org",
```
Did we mean `circadiastudy@mgh.harvard.edu`?

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+    "baseWebUrl": ${baseWebUrl},
+    "irbPassword": ${irbPassword},
+    "plusCodePrecision": "MEDIUM",
+    "shareParticipantLocation": true
```
```suggestion
    "shareParticipantLocation": false
```

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+  "statistics": [],
+
+  "client": {
+    "name": "circadia-angular-client",
```
`name` is actually deprecated. See `study-builder/studies/study-example.conf` file.

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+    },
+  ],
+
+  "activityTimestamp": "2015-01-01T00:00:00Z",
```
Circadia is not a migrated study, so we don't need `activityTimestamp`.

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+      "preconditionExpr": """
+      (
+(
+!user.studies["circadia"].forms["PREQUAL"].questions["ATYPICAL_CIRCADIAN_RHYTHM"].answers.hasOption("ACR_NO")
```
Option stable ids only need to be unique within the question itself. So it's okay to have option ids like `YES` and `NO` for multiple questions. You don't have to namespace your option ids with `ACR_`, `CRDS_`, etc. I recommend shortening the option ids -- it'll probably help make these PEX expressions shorter.

ScalaChamp replied:
@yufengwng Oh, I didn't know about that, thanks!

yufengwng commented on study-builder/studies/circadia/study.conf:
```diff
@@ -0,0 +1,672 @@
+        "type": "ACTIVITY_INSTANCE_CREATION",
+        "activityCode": "MINI_SLEEP_LOG"
+      },
+      "preconditionExpr": "true",
```
`preconditionExpr` and `cancelExpr` are optional. If we don't have a need for them, I recommend just removing it.

yufengwng commented on study-builder/studies/circadia/sendgrid-emails.conf:
```diff
@@ -0,0 +1,12 @@
+  "sendgridEmails": [
+    {
+      "key": "welcome",
+      "name": "CGC_WELCOME",
```
Is it `CGC` or `CRC`? Also, what does `CRC` stand for? I know `CGC` is probably Cancer Genomic Collective.

ScalaChamp replied:
@yufengwng `CRC` is  just shortened version of `CiRCadia`

aleksey-kuznetsov replied:
This email was replaced by one that is sent from Auth0 and it was removed [here](https://github.com/broadinstitute/ddp-study-server/pull/856).

yufengwng commented on study-builder/studies/circadia/emails/welcome.html:
```diff
@@ -0,0 +1,8 @@
+<html>
+<body>
+    <p>
+        -ddp.salutation-
```
I recommend using dynamic templates. See our [wiki page](https://broadinstitute.atlassian.net/wiki/spaces/DDP/pages/1182630029/Emails+Guide#Dynamic-Templates) for variables you can use (e.g. `{{salutation}}`).

aleksey-kuznetsov replied:
This email was a blank one when we started study development. It was removed and a new one was added [here](https://github.com/broadinstitute/ddp-study-server/pull/856).

yufengwng reviewed (commented):
Didn't read through all the files yet... a few comments to start. (Next time, I recommend breaking up your PR into multiple smaller PRs and getting them merged incrementally.)

ScalaChamp committed:
refactor configurations based on our standards, make it easier to understand and parse

aleksey-kuznetsov committed:
Circadia: fix auth0 issues

- Use `ASSETS_URL` in welcome email
- Add `ASSETS_URL` to config template
- Remove unnecessary invitation code from login page
- Extract `STUDY_EMAIL` to config
- Configure correct `from` field for welcome email

ScalaChamp committed:
DDP-6604 | fix bugs in Health and Lifestyle Questionnaire for Circadia

yufengwng commented on study-builder/config/circadia-config.json.ctmpl:
```diff
@@ -0,0 +1,38 @@
+  {{end}}
+    "BASE_URL": "{{$conf.baseWebUrl}}",
+    "LOGIN_DOMAIN": "{{$conf.auth0.loginDomain}}",
+    "ASSETS_URL": "{{$conf.assetsBucketName}}",
```
This one is a bit confusing -- we have a bucket name but we're saying it's a URL?

yufengwng reviewed (approved):
I think there's some cleanup to do and a bit of concern about the DSM types, but probably good to be merged in and be cleaned up as we go.

ScalaChamp committed:
Merge branch 'develop' into feature/circadia
